change(state): Add note subtree indexes for new and existing blocks

teor2345 opened the pull request:
## Motivation

This is needed to support "spend before sync" in lightwalletd.

Closes https://github.com/ZcashFoundation/zebra/issues/6953.

### Specifications

See https://github.com/zcash/zcash/pull/6677/files#diff-decae4be02fb8a47ab4557fe74a9cb853bdfa3ec0fa1b515c0a1e5de91f4ad0bR1458-R1475

The specs for subtrees aren't in a Zcash specification yet.

### Complex Code or Requirements

The upgrade runs concurrently with writing new blocks and subtrees, but the subtree indexes they write are different.

This code runs in separate threads, so we're careful to handle cancel, drop, and shutdowns.

Rebuilding subtrees could be CPU-intensive and take a long time.

## Solution

- Write subtrees to the finalized state for new blocks
- Upgrade existing finalized states to calculate and add subtrees
    - Assert that subtree indexes are correct
    - Log when each subtree is added during an upgrade, because upgrades are slow
- Bump the database format version to 25.2.0
- Check the subtree upgrade is valid after it has completed, and on every Zebra launch

Bug fixes:
- Prevent a concurrency bug by using the correct API (it's probably impossible anyway)

Refactors:
- Simplify subtree APIs
- Simplify subtree upgrade code
- Clarify some comments

### Testing

This PR has a state validity test. There are also asserts that subtrees are added in order and continuously.

## Review

This PR is needed to test the subtree RPC method in PR #7436.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Related Work

This is a copy of PR #7350 with some fixes to make it work with the latest `main` branch.

teor2345 committed:
Bump the database format version to 25.2.0

teor2345 committed:
Copy the add_subtrees upgrade from the original branch

teor2345 committed:
Copy the database write changes in shielded.rs from the original branch

teor2345 committed:
Copy the tree API changes from the original branch

teor2345 committed:
Simplify subtree APIs to avoid exposing frontiers

teor2345 committed:
Fix a dead code warning by re-using existing methods

teor2345 committed:
Use mpsc::Receiver<CancelFormatChange> in the subtree upgrade

teor2345 committed:
Run the subtree upgrade on startup

teor2345 commented on zebra-state/src/service/finalized_state/zebra_db/shielded.rs:
```diff
@@ -402,19 +399,32 @@ impl DiskWriteBatch {
-        //     self.zs_insert(&sapling_subtree_cf, subtree.index, subtree.into_data());
-        // }
+        if let Some(subtree) = trees.sapling_subtree {
+            self.insert_sapling_subtree(zebra_db, subtree);
```
Note: Since this PR writes a new format to the database, it can't merge until after the 1.2.0 release.

teor2345 committed:
Fix a confusing 'upgrade complete' log

teor2345 commented:
On testnet this upgrade takes less than 1 second:
```
2023-08-31T04:31:39.100268Z  INFO zebra_state::service::finalized_state::disk_format::upgrade: marked database format as upgraded running_version=Version { major: 25, minor: 2, patch: 0 } format_upgrade_version=Version { major: 25, minor: 1, patch: 1 } disk_version=Version { major: 25, minor: 1, patch: 0 }
2023-08-31T04:31:39.296963Z  INFO zebra_state::service::finalized_state::disk_format::upgrade: marked database format as upgraded running_version=Version { major: 25, minor: 2, patch: 0 } format_upgrade_version=Version { major: 25, minor: 2, patch: 0 } disk_version=Version { major: 25, minor: 1, patch: 1 }
2023-08-31T04:31:39.296892Z  INFO zebra_state::service::finalized_state::disk_format::upgrade: Zebra automatically upgraded the database format to: newer_running_version=Version { major: 25, minor: 2, patch: 0 }
```

Edit: This is probably because there are no subtrees. On mainnet the same upgrade panics because it's missing a Sapling subtree.

teor2345 committed:
Clarify some comments and error messages

teor2345 committed:
Simplify prev_tree unwrap to avoid an (impossible?) concurrency bug

teor2345 committed:
Use separate subtree writing functions

teor2345 committed:
Use common note commitment list code

teor2345 committed:
Fix subtree completion condition and add asserts

teor2345 committed:
Simplify subtree API and avoid exposing Address

teor2345 committed:
Fix API compatibility when Arcs are removed

teor2345 committed:
Log when each subtree is added

teor2345 committed:
If a format change is cancelled, don't mark the database as upgraded or do format checks

teor2345 committed:
Log subtree progress about once every two minutes

teor2345 commented:
I ended up doing a lot of fixes and refactors to this code to find two underlying bugs:
- the subtree loop condition was ignoring subtrees that were completed by the last note commitment in a block
- when an upgrade task was cancelled, the format checks and disk version bump could still run

On mainnet the upgrade takes about 1 second per subtree, I'm running it overnight locally so I can test the RPC.

arya2 committed:
Adds a state validity check for subtrees upgrade

upbqdn commented on zebra-chain/src/orchard/tree.rs:
```diff
@@ -347,28 +347,47 @@ impl NoteCommitmentTree {
+    /// Returns the subtree index at [`TRACKED_SUBTREE_HEIGHT`].
+    /// Returns `None` if the tree is empty.
+    #[allow(clippy::unwrap_in_result)]
+    pub fn subtree_index(&self) -> Option<NoteCommitmentSubtreeIndex> {
```
This method doesn't check if the frontier forms a complete subtree at `TRACKED_SUBTREE_HEIGHT`, we should mention this in the docs.

teor2345 replied:
Fixed in [e273588](https://github.com/ZcashFoundation/zebra/pull/7437/commits/e2735882f7f42eb3df03296a9546e9e2e4a0f7d4)

upbqdn commented on zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:
```diff
@@ -0,0 +1,221 @@
+//! Fully populate the Sapling and Orchard note commitment subtrees for existing blocks in the database.
```
###### Optional

```suggestion
//! Fully populates the Sapling and Orchard note commitment subtrees for existing blocks in the database.
```

teor2345 commented on zebra-chain/src/orchard/tree.rs:
```diff
@@ -347,28 +347,47 @@ impl NoteCommitmentTree {
-    /// Returns subtree address at [`TRACKED_SUBTREE_HEIGHT`]
-    pub fn subtree_address(tree: &NonEmptyFrontier<Node>) -> incrementalmerkletree::Address {
-        incrementalmerkletree::Address::above_position(
+    /// Returns the subtree index at [`TRACKED_SUBTREE_HEIGHT`].
```
```suggestion
    /// Returns the subtree index at [`TRACKED_SUBTREE_HEIGHT`].
    /// This is the number of complete or incomplete subtrees that are currently in the tree.
```

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:
```diff
@@ -0,0 +1,428 @@
+        if end_of_block_subtree_index.0 <= subtree_count && !tree.is_complete_subtree() {
+            prev_tree = Some(tree);
+            continue;
+        }
```
Consider moving this into an `else` clause at the end of the subtree creation code, to make the condition less confusing.

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:
```diff
@@ -0,0 +1,428 @@
+            continue;
+        }
+
+        if let Some((index, node)) = tree.completed_subtree_index_and_root() {
```
// The end of the block is also the leaf at the end of a subtree

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:
```diff
@@ -0,0 +1,428 @@
+                "trees are inserted in order with no gaps"
+            );
+            write_sapling_subtree(upgrade_db, index, height, node);
+        } else {
```
// The end of the block is in the next subtree

teor2345 commented:
> On mainnet the upgrade takes about 1 second per subtree, I'm running it overnight locally so I can test the RPC.

Here is the performance of the mainnet upgrade on my machine:
- ~1100 sapling subtrees: 15 minutes
- ~700 orchard subtrees: 5 minutes

So the total upgrade time is 20 minutes.

Here are the logs from the upgrade task:
```
2023-08-31T06:40:07.609922Z  INFO zebra_state::service::finalized_state::disk_format::upgrade: trying to open older database format: launching upgrade task running_version=Version { major: 25, minor: 2, patch: 0 } disk_version=Version { major: 25, minor: 1, patch: 1 }
2023-08-31T06:40:07.610006Z  INFO zebra_state::service::finalized_state::disk_db: the open file limit is high enough for Zebra current_limit=1024 min_limit=512 ideal_limit=1024
2023-08-31T06:40:07.846572Z  INFO zebra_state::service::finalized_state::disk_db: Opened Zebra state cache at /home/dev/.cache/zebra-custom/state/v25/mainnet
2023-08-31T06:40:07.848013Z  INFO zebra_state::service::finalized_state: loaded Zebra state cache tip=Some((Height(2210287), block::Hash("0000000000ec22974f7a26b974dcbaf095615e6e8a2d5b508b09d9cc3dd3c80c")))
2023-08-31T06:40:08.016508Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added sapling subtree height=Height(558822) index=0
2023-08-31T06:41:10.854368Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added sapling subtree height=Height(1733780) index=100
2023-08-31T06:42:54.680556Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added sapling subtree height=Height(1741676) index=200
2023-08-31T06:44:29.688948Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added sapling subtree height=Height(1753551) index=300
2023-08-31T06:45:44.096756Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added sapling subtree height=Height(1762369) index=400
2023-08-31T06:47:10.256812Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added sapling subtree height=Height(1771447) index=500
2023-08-31T06:48:27.590899Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added sapling subtree height=Height(1783656) index=600
2023-08-31T06:49:50.870770Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added sapling subtree height=Height(1791579) index=700
2023-08-31T06:51:26.138087Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added sapling subtree height=Height(1800361) index=800
2023-08-31T06:53:06.716403Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added sapling subtree height=Height(1809496) index=900
2023-08-31T06:54:29.509025Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added sapling subtree height=Height(1819275) index=1000
2023-08-31T06:55:23.494550Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added sapling subtree height=Height(2118375) index=1100
2023-08-31T06:55:24.233642Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added orchard subtree height=Height(1707429) index=0
2023-08-31T06:55:54.634851Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added orchard subtree height=Height(1836742) index=100
2023-08-31T06:56:35.204731Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added orchard subtree height=Height(1857929) index=200
2023-08-31T06:57:21.018305Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added orchard subtree height=Height(1881127) index=300
2023-08-31T06:58:03.194716Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added orchard subtree height=Height(1925537) index=400
2023-08-31T06:58:38.744790Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added orchard subtree height=Height(1957755) index=500
2023-08-31T06:59:12.502288Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added orchard subtree height=Height(2011784) index=600
2023-08-31T06:59:41.355723Z  INFO zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: calculated and added orchard subtree height=Height(2084076) index=700
2023-08-31T06:59:45.984703Z  INFO zebra_state::service::finalized_state::disk_format::upgrade: marked database format as upgraded running_version=Version { major: 25, minor: 2, patch: 0 } format_upgrade_version=Version { major: 25, minor: 2, patch: 0 } disk_version=Version { major: 25, minor: 1, patch: 1 }
2023-08-31T06:59:45.984716Z  INFO zebra_state::service::finalized_state::disk_format::upgrade: Zebra automatically upgraded the database format to: newer_running_version=Version { major: 25, minor: 2, patch: 0 }
```

teor2345 committed:
Orchard is faster, decrease log interval

arya2 committed:
updates subtree state validity check

teor2345 committed:
Clarify subtree index docs

arya2 commented on zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:
```diff
@@ -0,0 +1,436 @@
+/// Check that Orchard note commitment subtrees were correctly added.
+///
+/// Returns false if there are missing or bad subtrees.
+fn check_orchard_subtrees(db: &ZebraDb) -> bool {
```
This is equivalent to `check_sapling_subtrees` except all instances of "sapling" have been replaced with "orchard".

teor2345 committed:
Move a log to the correct location

teor2345 committed:
Refactor subtree upgrade to remove duplicate inverted loop conditions

teor2345 committed:
Add a subtree format check when there is no upgrade

teor2345 commented on zebra-state/src/service/finalized_state/zebra_db.rs:
```diff
@@ -108,9 +108,10 @@ impl ZebraDb {
+            // the database format should be valid.
             // (There's no format change here, so the format change checks won't run.)
             DbFormatChange::check_for_duplicate_trees(db.clone());
+            upgrade::add_subtrees::check(&db.clone());
```
#### Future Work

This check got skipped, maybe we could document that somewhere. (Or put all the checks in a single function, and call it from here.)

teor2345 committed:
Fix an off-by-one error with the final subtree check

teor2345 committed:
Use error-level logs for database format checks

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:
```diff
@@ -0,0 +1,454 @@
+///
+/// If a note commitment subtree is missing or incorrect.
+fn check_sapling_subtrees(db: &ZebraDb) -> bool {
+    let Some(NoteCommitmentSubtreeIndex(mut first_incomplete_subtree_index)) =
+        db.sapling_tree().subtree_index()
+    else {
+        return true;
+    };
+
+    // If there are no incomplete subtrees in the tree, also expect a subtree for the final index.
+    if db.sapling_tree().is_complete_subtree() {
+        first_incomplete_subtree_index += 1;
+    }
+
+    let mut is_valid = true;
+    for index in 0..first_incomplete_subtree_index {
```
There was an off-by-one error in this check: if the current tree is also a complete subtree, we need to check that index as well.

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:
```diff
@@ -0,0 +1,454 @@
+    for index in 0..first_incomplete_subtree_index {
+        // Check that there's a continuous range of subtrees from index [0, first_incomplete_subtree_index)
+        let Some(subtree) = db.sapling_subtree_by_index(index) else {
+            error!(index, "missing subtree");
```
I changed the logs to errors because it's unrecoverable, so we want it big and red.

teor2345 reviewed (commented):
When testing this code locally, I'm getting this panic:
```
service::finalized_state::zebra_db::block::tests::vectors::test_block_db_round_trip
The application panicked (crashed).
Message:  Sapling note commitment trees must exist for all heights below the finalized tip
Location: zebra-state/src/service/finalized_state/zebra_db/shielded.rs:156
```

This is a bug in the test, because that assertion isn't true any more after #7239.

teor2345 committed:
Merge branch 'main' into subtree-upgrade-data-only

teor2345 commented:
I've finished with test coverage fixes now, feel free to have a go at the panic bug. (I wonder if it can happen on `main` and needs a fix before we tag the release?)

teor2345 committed:
Skip format checks in tests that create invalid formats

teor2345 commented:
> When testing this code locally, I'm getting this panic:
> 
> ```
> service::finalized_state::zebra_db::block::tests::vectors::test_block_db_round_trip
> The application panicked (crashed).
> Message:  Sapling note commitment trees must exist for all heights below the finalized tip
> Location: zebra-state/src/service/finalized_state/zebra_db/shielded.rs:156
> ```
> 
> This is a bug in the test, because that assertion isn't true any more after #7239.

Actually, it's a bug in this PR, which breaks the test. This PR is running format checks during the tests, and some tests produce invalid formats. I fixed the test so it skips the format upgrade & checks in 44ac0035f8.

teor2345 commented:
This doesn't impact the checks in the `main` branch, because they look up a range of trees, and the test doesn't add any trees. (So the check passes because the state looks "empty".)

arya2 committed:
fix state validity test

upbqdn commented on zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:
```diff
@@ -0,0 +1,466 @@
+}
+
+/// Writes a Orchard note commitment subtree to `upgrade_db`.
+fn write_orchard_subtree(
```
This method inserts a root of a subtree, not a subtree. We started calling them subtrees everywhere, but they're, in fact, only roots. This will likely cause a naming issue if we ever need to store actual subtrees.

mpguerra replied:
this looks like the sort of information that gets lost and causes misunderstandings in future :) Can we rename the methods and update comments to correctly reflect what we are storing? Since we want this PR to merge soon we should do it in a follow up PR

upbqdn replied:
However, ECC uses the same naming.

teor2345 replied:
This is now ticket #7463

teor2345 commented:
Failing due to a permissions error, which could be permanent if it's caused by other DevOps deployments to Google Cloud:
> Run gcloud compute instances create-with-container "sync-past-checkpoint-7437-merge-85fd613" \
ERROR: (gcloud.compute.instances.create-with-container) Could not fetch resource:
> - Required 'compute.subnetworks.use' permission for 'projects/zfnd-dev-net-spoke-0/regions/us-east1/subnetworks/dev-default-ue1'

https://github.com/ZcashFoundation/zebra/actions/runs/6052445578/job/16427401749?pr=7437#step:11:80

@gustavovalverde any idea what's going on here?

teor2345 commented:
> Failing due to a permissions error, which could be permanent if it's caused by other DevOps deployments to Google Cloud:
> 
> > Run gcloud compute instances create-with-container "sync-past-checkpoint-7437-merge-85fd613" 
> > ERROR: (gcloud.compute.instances.create-with-container) Could not fetch resource:
> > 
> > * Required 'compute.subnetworks.use' permission for 'projects/zfnd-dev-net-spoke-0/regions/us-east1/subnetworks/dev-default-ue1'
> 
> [ZcashFoundation/zebra/actions/runs/6052445578/job/16427401749?pr=7437#step:11:80](https://github.com/ZcashFoundation/zebra/actions/runs/6052445578/job/16427401749?pr=7437#step:11:80)
> 
> @gustavovalverde any idea what's going on here?

This seems to have been a temporary error.

teor2345 commented:
@arya2 or @upbqdn, just a reminder that one of you needs to approve this PR before it can merge. (Edit: CI has passed.)

teor2345 commented:
@arya2 would you mind opening a PR with comments that explain the second part of the subtree validity checks? I want to make sure we know what's happening when we come back to that code later.

But that isn't a blocker for merging this PR, and I'd like to get this PR merged so we can write more tests.

teor2345 committed:
Merge branch 'main' into subtree-upgrade-data-only

Manual merge of both changes to zebra-chain/src/subtree.rs

teor2345 committed:
Add a concurrency comment to subtree by height methods

teor2345 committed:
Add individual subtree state methods: reverts removing these methods in an earlier PR

teor2345 committed:
fastmod "subtrees_by_index" "subtree_list_by_index_for_rpc"

upbqdn reviewed (approved):
utACK

teor2345 commented:
@mergifyio refresh

teor2345 commented:
Failed in the merge PR in the `non_blocking_logger` test with an unexpected hang:
https://github.com/ZcashFoundation/zebra/actions/runs/6078117140/job/16488853272#step:15:6963

We might need to increase the timeout for that test.

teor2345 commented:
GitHub actions seems to be having some issues, I've bumped this PR up to critical priority so it merges first.

teor2345 commented:
@mergifyio refresh

teor2345 commented:
@mergifyio refresh

teor2345 commented:
@mergifyio refresh

teor2345 commented:
GitHub said there was an API request issue which got resolved:
https://www.githubstatus.com/history

upbqdn commented:
@mergifyio refresh
